dxr/query.py

```diff
@@ -126,52 +161,39 @@ def group_filters_by_name(predicate):
-                    'filter': {
-                        'and': ors
-                    }
-                }
+        return self._do_search_and_highlight(_make_es_query(filters, is_line_query),
+                                             filters, is_line_query,
+                                             offset, limit)
+
+    # Test: If var-ref (or any structural query) returns 2 refs on one line, they should both get highlit.
+
+    def promoted_paths(self, term, promote_limit=5):
+        """Return a tuple (total number of path results, promoted paths, querystring),
+        where the number of promoted paths is no more than promote_limit."""
+
+        # Filter out the FILE domain filters, and add our own path filter at the end.
+        file_filters = [[f for f in named_filters if f.domain == FILE] for named_filters in
+                        self._filters] + [[PathFilter(term, self.enabled_plugins)]]
+        filtered_query = _make_es_query(file_filters, False)
+        path_query = {'constant_score': {'query': filtered_query, 'boost': 0.5}}
+        # We add a second query that boosts exact path segment matches.
+        match_field = 'path.segments' if term['case_sensitive'] else 'path.segments_lower'
+        exact_query = {
+            'filtered': {
+                'query': {'constant_score': {'query': {'match': {match_field: term['arg']}},
+                                             'boost': 2.0}},
+                # Borrow the same filters the path query uses.
+                'filter': filtered_query['filtered']['filter']
             }
-        else:
-            query = {
-                'match_all': {}
+        }
+        query = {
+            'dis_max': {
+                'queries': [path_query, exact_query]
             }
-
-        results = self.es_search(
-            {'query': query,
-             'sort': ['path', 'number'] if is_line_query else ['path'],
-             'from': offset,
-             'size': limit},
-            doc_type=LINE if is_line_query else FILE)['hits']
-        result_count = results['total']
-        results = [r['_source'] for r in results['hits']]
-
-        path_highlighters = [f.highlight_path for f in chain.from_iterable(filters)
-                             if hasattr(f, 'highlight_path')]
-        return {'result_count': result_count,
-                'results': self._line_query_results(filters, results, path_highlighters)
-                           if is_line_query
-                           else self._file_query_results(results, path_highlighters)}
-
-        # Test: If var-ref (or any structural query) returns 2 refs on one line, they should both get highlit.
+        }
+        # We ask not to sort by path and line, to use score ranking instead.
+        result_count, results = self._do_search_and_highlight(query, file_filters, False, 0, promote_limit, False)
+        return result_count, results, ' '.join((str(f) for f in chain.from_iterable(file_filters)))
```

Comment:
str(), not unicode()? What if somebody searches for something unicodey? If I type "readmé" in the search field, it crashes. (Add a test, too.)